zenoh/src/net/routing/interceptor/downsampling.rs

```diff
@@ -0,0 +1,160 @@
+//!
+//! [Click here for Zenoh's documentation](../zenoh/index.html)
+
+use crate::net::routing::interceptor::*;
+use std::sync::{Arc, Mutex};
+use zenoh_config::DownsamplingItemConf;
+use zenoh_protocol::core::key_expr::OwnedKeyExpr;
+
+const RATELIMIT_STRATEGY: &str = "ratelimit";
+
+// EgressMsgDownsamplerRatelimit implements default ratelimit strategy, it has following
+// configuration parameters:
+//   threshold  - mandatory, indicates the minimum interval between messages.
+//                All messages arriving with a smaller interval will be skipped
+//   keyexprs   - optional, list of key expressions, only messages matching this expression will be processed,
+//                the rest will be passed as is.
+//   interfaces - optional, list of interfaces messages will be processed on,
+//                the rest will be passed as is.
+//
+// TODO: we can also add decimation strategy, with mandatory "factor" parameter
+
+pub(crate) struct EgressMsgDownsamplerRatelimit {
+    keyexprs: Option<Vec<OwnedKeyExpr>>,
+    threshold: std::time::Duration,
+    latest_message_timestamp: Arc<Mutex<std::time::Instant>>,
+}
+
+impl InterceptorTrait for EgressMsgDownsamplerRatelimit {
+    fn intercept(
+        &self,
+        ctx: RoutingContext<NetworkMessage>,
+    ) -> Option<RoutingContext<NetworkMessage>> {
+        if let Some(cfg_keyexprs) = self.keyexprs.as_ref() {
+            let mut matched = false;
+            if let Some(keyexpr) = ctx.full_key_expr() {
+                for cfg_keyexpr in cfg_keyexprs {
+                    if cfg_keyexpr.intersects(&keyexpr) {
+                        matched = true;
+                    }
+                }
+            }
+            if !matched {
+                return Some(ctx);
+            }
+        }
+
+        let timestamp = std::time::Instant::now();
+        let mut latest_message_timestamp = self.latest_message_timestamp.lock().unwrap();
+
+        if timestamp - *latest_message_timestamp >= self.threshold {
+            *latest_message_timestamp = timestamp;
+            log::debug!("Interceptor: Passed threshold, passing.");
+            Some(ctx)
+        } else {
+            log::debug!("Interceptor: Skipped due to threshold.");
+            None
+        }
+    }
+}
+
+impl EgressMsgDownsamplerRatelimit {
+    pub fn new(conf: DownsamplingItemConf) -> Self {
+        if let Some(threshold_ms) = conf.threshold_ms {
+            let threshold = std::time::Duration::from_millis(threshold_ms);
+            Self {
+                keyexprs: conf.keyexprs,
+                threshold,
+                // TODO (sashacmc): I need just := 0, but how???
+                latest_message_timestamp: Arc::new(Mutex::new(
+                    std::time::Instant::now() - threshold,
+                )),
+            }
+        } else {
+            // TODO (sashacmc): how correctly process an error?
+            panic!("Rate limit downsampler shoud have a threshold_ms parameter");
```

Comment:
How better process error from the interceptor?
Do we have some policies for it?

Reply:
In general, instead of returning `Self` you should return a `ZResult<Self>`, by doing so you can return an error in case of wrong parameters.

For this specific case, I wonder why `ratelimit` is an `Option` instead of being mandatory. Maybe we should separate the parsing of the config from the initialization of the object.

Reply:
As many checks as possible should be done at the factory instantiation.
Ultimately the `interceptor_factories` function should indeed return a `ZResult` as well as all the stack leading to it so that we might return an error to `zenoh::open`.

Reply:
`ratelimit` is optional because it depends on downsampling strategy for the ratelimit strategy it is mandatory, for other may be not.

I cannot agree that all checks should be done in interceptor_factories; on the contrary, I would prefer to encapsulate all checks inside specific interceptors.
But I get your point with the error processing and I will rework configuration processing, thanks!